cpp/src/arrow/compute/kernels/codegen_internal.h

```diff
@@ -485,18 +502,44 @@ struct ScalarUnaryNotNullStateful {
   struct ArrayExec<Type, enable_if_base_binary<Type>> {
     static void Exec(const ThisType& functor, KernelContext* ctx, const ExecBatch& batch,
                      Datum* out) {
-      typename TypeTraits<Type>::BuilderType builder;
-      VisitArrayValuesInline<Arg0Type>(*batch[0].array(), [&](util::optional<ARG0> v) {
-        if (v.has_value()) {
-          KERNEL_RETURN_IF_ERROR(ctx, builder.Append(functor.op.Call(ctx, *v)));
-        } else {
-          KERNEL_RETURN_IF_ERROR(ctx, builder.AppendNull());
+      // TODO: This code path is currently inadequately tested.
+
+      using offset_type = typename Type::offset_type;
```

Comment:
I'm not convinced why this is doing everything by hand instead of relying on the array builder as before.

Reply:
The validity bitmap is already precomputed (zero-copied, in fact) by the executor so using the builder natively here is computationally wasteful

Reply:
I'm reverting this change since no kernel uses it and adding a comment explaining that it is suboptimal.